ASN1 Fuzzer

dannyallover opened the pull request:
I've developed a protobuf to represent an ASN.1 PDU defined in spec X.690 (2015). Along with the converter and LPM, the protobuf can be used to generate interesting inputs for a fuzzer and potentially increase coverage.

I hope to establish a proto directory where future protos can live. By placing my proto and converter here, I hope that other projects that deal with ASN.1 can use this to improve their fuzzers.

dannyallover commented:
PTAL @sleevi @Dor1s

sleevi commented on proto/asn1-pdu-proto/README.md:
```diff
@@ -0,0 +1,5 @@
+# ASN.1 PDU Protobuf
+This protobuf represents an ASN.1 PDU defined in spec X.690 (2015). Along with the converter and LPM, the protobuf can be used to generate intersting inputs for a fuzzer.
```
```suggestion
This [protobuf](https://developers.google.com/protocol-buffers) represents a
DER-encoded ASN.1 PDU, defined in X.690 (2015). When used with
[libprotobuf-mutator](https://github.com/google/libprotobuf-mutator), the protobuf can be used
to produce valid DER sequences, as well as syntactically invalid (e.g. BER) and structurally
invalid (e.g. random data) inputs for a fuzzer.
```

The motivation behind these changes are:
* The Protobuf largely reflects the DER-encoding. There are other encodings (e.g. XER, CER, PER)
   which this doesn't fully touch on, and this isn't a full X.680 AST.
* This is Markdown, so we can linkify things like LPM :) 
* Calling out some of the capabilities here - it's DER, sometimes BER (e.g. indefinite encodings),
  and sometimes just invalid (e.g. the random data)
* Wrap to 78/80

sleevi commented on proto/asn1-pdu-proto/asn1_proto_to_der.cc:
```diff
@@ -0,0 +1,158 @@
+
+void ASN1ProtoToDER::EncodeValue(const Value& val) {
+  for (const auto& val_ele : val.val_array()) {
+    if (!kErrRecursionLimitReached) {
+      if (val_ele.has_pdu()) {
+        EncodePDU(val_ele.pdu());
+      } else {
+        encoder_.insert(encoder_.end(), val_ele.val_bits().begin(),
+                        val_ele.val_bits().end());
+      }
+    }
```
```suggestion
    if (kErrRecursionLimitReached) {
      // If the message exceeds the recursion limit, abort processing the
      // protobuf in order to limit uninteresting work.
      return;
    }
    if (val_ele.has_pdu()) {
      EncodePDU(val_ele.pdu());
    } else {
      encoder_.insert(encoder_.end(), val_ele.val_bits().begin(),
                      val_ele.val_bits().end());
    }
```

In general, try to handle errors early and avoid introducing the horizontal depth. Each nesting level is another piece of data a reader has to keep on their mental stack, to try to remember what the conditions are for the current statement. Particularly with `if` statements, the reader has to read all the way to end of the closing brace to determine whether there is an `else` statement that might processing things differently.

By using early returns, you make it a little easier to reason about the state. Particularly for error handling, the idea is "handle the exceptional situation as appropriate, as early as possible", such as returning an error code or taking some corrective action, so that the reader can continue thinking about what "normal" processing looks like.

dannyallover replied:
Very insightful and accurate -- change accepted!

sleevi commented on proto/asn1-pdu-proto/asn1_proto_to_der.cc:
```suggestion
  size_t len_pos = encoder_.size();
  EncodeValue(pdu.val());
  EncodeLength(pdu.len(), encoder_.size() - len_pos, len_pos);
```

This might be a little easier to read, mostly because the `- id_size` and `+ id_size` made it a little harder to mentally track what the sizes were, and the subtlety between `value_size` and `size_before_insertion + id_size` is took a bit to process. What do you think of the above?

dannyallover replied:
Yes, I agree. It improves readability, especially since there's less to track here. Change accepted!

sleevi commented on proto/asn1-pdu-proto/asn1_proto_to_der.cc:
I'd say this is fairly obvious, and so probably doesn't need a comment, although if you do comment, probably something as simple as.

```suggestion
  // Reset the previous state.
```

sleevi commented on proto/asn1-pdu-proto/asn1_proto_to_der.cc:
```suggestion
  kErrRecursionLimitReached = false;

  EncodePDU(pdu);
```

Here, we use vertical whitespace to distinguish between different "tasks" of this method - initialization vs execution

sleevi commented on proto/asn1-pdu-proto/asn1_proto_to_der.h:
```diff
@@ -0,0 +1,74 @@
+#ifndef ASN1_PROTO_TO_DER_H_
+#define ASN1_PROTO_TO_DER_H_
+
+#include <cmath>
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <typeinfo>
+#include <vector>
```
```suggestion
#include <stdint.h>

#include <string>
#include <vector>

```

Following with https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes , we place C headers, then C++ headers, then your project's .h , each with newline separations.

We remove the unused includes.

sleevi commented on proto/asn1-pdu-proto/asn1_proto_to_der.h:
```diff
@@ -0,0 +1,74 @@
+  void EncodeLength(const Length& len, size_t actual_len, size_t len_pos);
+
+  // Writes |raw_len| to |encoder_| at |len_pos|.
+  void EncodeOverrideLength(const std::string raw_len, const size_t len_pos);
```
```suggestion
  void EncodeOverrideLength(const std::string& raw_len, const size_t len_pos);
```

And update the corresponding implementation. This avoids forcing a copy of raw_len to be made

Dor1s commented on proto/asn1-pdu-proto/asn1_proto_to_der.cc:
is it correct to increase the `depth_` before this check? e.g. let's say `kRecursionLimit` is `0`, so we'll never execute this `EncodePDU`, but the value `depth_` we will have is going to be `1`?

dannyallover replied:
Good catch! Change made :)

Dor1s commented:
@dannyallover looks like you've authored commits using different emails and that might be why the CLA bot is complaining:

```
$ git log -n 15 | egrep -i author | sort | uniq
Author: Danny Halawi <dannyhalawi15@gmail.com>
Author: Danny Halawi <dhalawi@google.com>
    Co-authored-by: sleevi <ryan.sleevi@gmail.com>
```

you do need to sign the CLA, but please try adding both email to your GitHub account, IIRC that should fix the issue.

Dor1s commented:
@googlebot what do you think now?

Dor1s commented:
@googlebot I fixed it.

dannyallover commented:
PTAL @Dor1s @sleevi I've addressed the comments + fixed final bugs + confirmed from my end that it's producing valid DER. I'm still getting the "need the CLA" after I added my second email to my github. I'll look into this today.

sleevi commented on proto/asn1-pdu/asn1_pdu_to_der.cc:
```diff
@@ -0,0 +1,159 @@
+// fail.
+static constexpr size_t kRecursionLimit = 200;
+
+// Signal that the recursion limit has been exceeded by a
+// message or value and to abort further processing.
+static bool kErrRecursionLimitReached = false;
```
Sorry I had missed this when you changed to a bool and treated it as a sentinel value, rather than a return value.

This should be a class member, like depth_. The original design had this as a static constant (like `#define` but cleaner), but your new design has it as a static global, which we [discourage](https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables)

For example, if you had two instances of this fuzzer running (... not like you normally would), they'd share the same state for tracking recursion depth. Making this part of the class should be fine here, as noted in the `.h` file.

sleevi commented on proto/asn1-pdu/asn1_pdu_to_der.cc:
```diff
@@ -0,0 +1,159 @@
+uint8_t ASN1PDUToDER::GetVariableIntLen(const uint64_t value,
+                                          const size_t base) {
+  uint8_t base_bits = log2(base);
+  for (uint8_t num_bits = (sizeof(value) - 1) * __CHAR_BIT__;
```
`__CHAR_BIT__` isn't the same as `CHAR_BIT`. You want the latter, which you get by

`#include <climits>`

See https://en.cppreference.com/w/cpp/types/climits for handy reference :)

(And throughout the file)

sleevi commented on proto/asn1-pdu/asn1_pdu_to_der.cc:
```diff
@@ -0,0 +1,159 @@
+  for (uint8_t num_bits = (sizeof(value) - 1) * __CHAR_BIT__;
+       num_bits >= base_bits; num_bits -= base_bits) {
+    if (value >> num_bits) {
+      return ceil((double)num_bits / base_bits) + 1;
```
Why the conversion to double here? That raises some concerns. Was this a compiler warning? Was it related to integer truncation? If it is needed for that reason, https://google.github.io/styleguide/cppguide.html#Casting is the guidance here :)

dannyallover replied:
@sleevi double was needed because num_bits/base_bits was rounding down to the nearest integer before it got piped into `ceil`.

sleevi commented on proto/asn1-pdu/asn1_pdu_to_der.cc:
```diff
@@ -0,0 +1,159 @@
+static bool kErrRecursionLimitReached = false;
+
+uint8_t ASN1PDUToDER::GetVariableIntLen(const uint64_t value,
+                                          const size_t base) {
```
You'll need to reformat this file - looks like it got misaligned by two spaces?

sleevi commented on proto/asn1-pdu/asn1_pdu_to_der.h:
```suggestion

  // Whether |depth_| exceeded the recursion limit, causing an early return.
  bool recursion_exceeded_;
};
```

sleevi commented on proto/asn1-pdu/asn1_pdu_to_der.cc:
```suggestion
  }
  ++depth_;
```

There's no need to "else" after a return, so helps avoid a layer of brace depth :)

sleevi commented:
@googlebot I consent.
